File: src/leetcode/0733.FloodFill/01/floodFill.java
import java.util.LinkedList;
import java.util.Queue;

public class Solution {
    // 从（sr,sc) 开始上色  row col
    // bfs
    public int[][] floodFill(int[][] image, int sr, int sc, int newColor) {
        int[][] dir = {{-1,0},{1,0},{0,-1},{0,1}};
        int srcColor = image[sr][sc];
        if(srcColor == newColor) {
            return image;
        }
        int m= image.length,n = image[0].length;
        Queue<int[]> queue = new LinkedList<int[]>();
        queue.offer(new int[]{sr,sc});
        image[sr][sc] = newColor;
        while(!queue.isEmpty()) {
            int[] cell = queue.poll();
            int x=cell[0],y=cell[1];
            for (int i=0;i<4;i++){
                int mx =  x+dir[i][0],my=y+dir[i][1];
                if(mx>=0 && mx<m && my>=0 && my<n) {
                    queue.offer(new int[]{mx,my});
                    image[mx][my] = newColor;
                }
            }
        }
        return image;
    }
}

File: leetcode/0733.FloodFill/floodFill.go
package lt733


/*
二维数组表示图像   值表示像素值大小
先给定一个坐标，以及颜色值，将像素值与起始坐标相同的点都渲染成新的颜色

寻址问题  BFS  DFS

*/
var (
    dx = []int{1, 0, 0, -1}
    dy = []int{0, 1, -1, 0}
)//四个方向 delta变量
func floodFill(image [][]int, sr int, sc int, newColor int) [][]int {

	//DFS
	curColor := image[sr][sc]
	if curColor != newColor {
		dfs(image, sr,sc, curColor,newColor)
	}
	return image
}

func dfs(image [][]int,  x,y,color,newColor int){
	if image[x][y] == color {
		image[x][y] = newColor
		//上下左右遍历
		for i:=0;i<4;i++{
			mx,my := x+dx[i], y+dy[i]
			if mx>=0 && mx <len(image) && my>=0 && my<len(image[0]) {
				dfs(image, mx,my,color,newColor)
			}
		}
	}
}